File: lib/games/crazy_eights/crazy_eights_engine.dart
import 'dart:math';

import 'card.dart';

enum CrazyEightsPhase { lobby, playing, gameOver }

class CrazyEightsPlayer {
  CrazyEightsPlayer({required this.id, required this.name, required this.isHost});
  final String id;
  final String name;
  final bool isHost;
  final List<Card> hand = [];
}

/// Classic Crazy Eights:
///   - Match top card by suit or rank
///   - 8s are wild — player picks new active suit
///   - If you can't play, draw one card (if it's playable, you may play it)
///   - First empty hand wins
class CrazyEightsEngine {
  CrazyEightsEngine({Random? rng}) : _rng = rng ?? Random();

  final Random _rng;
  final Map<String, CrazyEightsPlayer> players = {};
  final List<String> _order = [];
  CrazyEightsPhase phase = CrazyEightsPhase.lobby;

  final List<Card> drawPile = [];
  final List<Card> discardPile = [];
  Suit? activeSuit; // overrides top card suit when an 8 is played
  int currentIndex = 0;
  bool justDrew = false;
  String? winnerId;
  String? lastEvent;

  CrazyEightsPlayer? get current =>
      _order.isEmpty ? null : players[_order[currentIndex]];

  Card? get topCard => discardPile.isEmpty ? null : discardPile.last;

  Suit get activeOrTopSuit => activeSuit ?? topCard!.suit;

  // ---- Lobby ----
  CrazyEightsPlayer addPlayer({required String id, required String name, bool isHost = false}) {
    final p = CrazyEightsPlayer(id: id, name: name, isHost: isHost);
    players[id] = p;
    return p;
  }

  void removePlayer(String id) {
    if (phase != CrazyEightsPhase.lobby) return;
    players.remove(id);
  }

  bool get canStart =>
      phase == CrazyEightsPhase.lobby && players.length >= 2 && players.length <= 8;

  void start() {
    if (!canStart) return;
    drawPile.clear();
    discardPile.clear();
    activeSuit = null;
    drawPile.addAll(standardDeck()..shuffle(_rng));

    final dealCount = players.length == 2 ? 7 : 5;
    _order
      ..clear()
      ..addAll(players.keys.toList()..shuffle(_rng));
    for (var i = 0; i < dealCount; i++) {
      for (final pid in _order) {
        players[pid]!.hand.add(drawPile.removeLast());
      }
    }

    // Flip first non-8 card to start. Eights start the active suit fresh.
    while (drawPile.isNotEmpty) {
      final c = drawPile.removeLast();
      discardPile.add(c);
      if (c.rank != 8) break;
    }
    currentIndex = 0;
    justDrew = false;
    phase = CrazyEightsPhase.playing;
    lastEvent = '${current!.name} starts';
  }

  // ---- Play / draw ----
  bool canPlay(Card c) {
    final top = topCard;
    if (top == null) return true;
    if (c.rank == 8) return true;
    if (c.suit == activeOrTopSuit) return true;
    if (c.rank == top.rank) return true;
    return false;
  }

  /// Returns null on success, an error message on failure.
  String? playCard(String playerId, Card card, {Suit? declaredSuit}) {
    if (phase != CrazyEightsPhase.playing) return 'not playing';
    final p = players[playerId];
    if (p == null) return 'unknown player';
    if (p.id != current!.id) return 'not your turn';
    final inHand = p.hand.firstWhere(
        (c) => c == card,
        orElse: () => const Card(Suit.clubs, 0));
    if (inHand.rank == 0) return 'card not in hand';
    if (!canPlay(card)) return 'card does not match';
    if (card.rank == 8 && declaredSuit == null) return 'must declare a suit';

    p.hand.remove(card);
    discardPile.add(card);
    activeSuit = card.rank == 8 ? declaredSuit : null;
    justDrew = false;
    lastEvent = card.rank == 8
        ? '${p.name} played $card → ${declaredSuit!.glyph}'
        : '${p.name} played $card';

    if (p.hand.isEmpty) {
      phase = CrazyEightsPhase.gameOver;
      winnerId = p.id;
      return null;
    }
    _advanceTurn();
    return null;
  }

  /// Draws one card. If the drawn card is playable, the player may play it
  /// (their turn doesn't end yet). If not, turn passes.
  Card? drawOne(String playerId) {
    if (phase != CrazyEightsPhase.playing) return null;
    final p = players[playerId];
    if (p == null || p.id != current!.id) return null;
    if (drawPile.isEmpty) _reshuffle();
    if (drawPile.isEmpty) {
      // no cards available at all — pass
      _advanceTurn();
      return null;
    }
    final c = drawPile.removeLast();
    p.hand.add(c);
    lastEvent = '${p.name} drew a card';
    if (canPlay(c)) {
      justDrew = true;
      return c;
    }
    justDrew = false;
    _advanceTurn();
    return c;
  }

  /// Pass after a draw if the drawn card is unplayable or you choose not to.
  void passAfterDraw(String playerId) {
    final p = players[playerId];
    if (p == null || p.id != current!.id) return;
    if (!justDrew) return;
    justDrew = false;
    lastEvent = '${p.name} passed';
    _advanceTurn();
  }

  void _advanceTurn() {
    currentIndex = (currentIndex + 1) % _order.length;
    justDrew = false;
  }

  void _reshuffle() {
    if (discardPile.length <= 1) return;
    final top = discardPile.removeLast();
    drawPile.addAll(discardPile..shuffle(_rng));
    discardPile
      ..clear()
      ..add(top);
  }

  void reset() {
    phase = CrazyEightsPhase.lobby;
    drawPile.clear();
    discardPile.clear();
    activeSuit = null;
    currentIndex = 0;
    justDrew = false;
    winnerId = null;
    lastEvent = null;
    for (final p in players.values) {
      p.hand.clear();
    }
  }
}

File: android/app/src/main/java/com/example/jamboree/games/crazyeights/CrazyEightsEngine.kt
package com.example.ubapp.games.crazyeights

import kotlin.random.Random

enum class Suit(val short: String, val glyph: String) {
    CLUBS("C", "♣"), DIAMONDS("D", "♦"), HEARTS("H", "♥"), SPADES("S", "♠");
    val isRed: Boolean get() = this == DIAMONDS || this == HEARTS
}

/** rank 2..14 (J=11, Q=12, K=13, A=14). */
data class Card(val suit: Suit, val rank: Int) {
    val rankShort: String get() = when (rank) { 11 -> "J"; 12 -> "Q"; 13 -> "K"; 14 -> "A"; else -> "$rank" }
    val id: String get() = "${suit.short}$rank"
    override fun toString() = "$rankShort${suit.glyph}"
}

fun standardDeck(): List<Card> =
    Suit.entries.flatMap { s -> (2..14).map { Card(s, it) } }

enum class CrazyEightsPhase { LOBBY, PLAYING, GAME_OVER }

/** Host-configurable house rules. Defaults reproduce the classic game. */
data class CrazyEightsOptions(
    val startingHandSize: Int? = null,
    val jackSkips: Boolean = false,
    val queenReverses: Boolean = false,
    /** Playing a 2 forces the next player to draw two cards and lose their turn. */
    val twosDrawTwo: Boolean = false,
)

class CrazyEightsPlayer(val id: String, val name: String, val isHost: Boolean) {
    val hand: MutableList<Card> = mutableListOf()
}

/**
 * Classic Crazy Eights:
 *  - Match top card by suit or rank
 *  - 8s are wild — player picks new active suit
 *  - If you can't play, draw one card (if it's playable, you may play it)
 *  - First empty hand wins
 */
class CrazyEightsEngine(private val rng: Random = Random.Default) {
    val tutorialVote = com.example.ubapp.tutorials.TutorialVote()
    val players: MutableMap<String, CrazyEightsPlayer> = linkedMapOf()
    private val order: MutableList<String> = mutableListOf()
    var phase: CrazyEightsPhase = CrazyEightsPhase.LOBBY
    var options: CrazyEightsOptions = CrazyEightsOptions()
        private set

    val drawPile: MutableList<Card> = mutableListOf()
    val discardPile: MutableList<Card> = mutableListOf()
    /** Overrides top card suit when an 8 is played. */
    var activeSuit: Suit? = null
    var currentIndex: Int = 0
    var justDrew: Boolean = false
    private var direction: Int = 1
    var winnerId: String? = null
    var lastEvent: String? = null
    /** Consecutive turns where the draw pile is exhausted and can't be
     *  replenished and the player can't act. A full lap of these ends the
     *  game so a blocked board doesn't loop forever. */
    private var stalePasses = 0

    val current: CrazyEightsPlayer? get() = if (order.isEmpty()) null else players[order[currentIndex]]
    val topCard: Card? get() = discardPile.lastOrNull()
    val activeOrTopSuit: Suit get() = activeSuit ?: topCard!!.suit

    fun addPlayer(id: String, name: String, isHost: Boolean = false): CrazyEightsPlayer {
        val p = CrazyEightsPlayer(id, name, isHost); players[id] = p; return p
    }
    fun removePlayer(id: String) { if (phase == CrazyEightsPhase.LOBBY) players.remove(id) }
    val canStart: Boolean get() = phase == CrazyEightsPhase.LOBBY && players.size in 2..8

    fun setOptions(o: CrazyEightsOptions) {
        if (phase != CrazyEightsPhase.LOBBY) return
        options = o.copy(startingHandSize = o.startingHandSize?.coerceIn(3, 10))
    }

    fun start() {
        if (!canStart) return
        drawPile.clear(); discardPile.clear(); activeSuit = null; direction = 1
        drawPile.addAll(standardDeck().shuffled(rng))
        val dealCount = options.startingHandSize ?: (if (players.size == 2) 7 else 5)
        stalePasses = 0
        order.clear(); order.addAll(players.keys.shuffled(rng))
        repeat(dealCount) {
            for (pid in order) players[pid]!!.hand.add(drawPile.removeAt(drawPile.size - 1))
        }
        while (drawPile.isNotEmpty()) {
            val c = drawPile.removeAt(drawPile.size - 1)
            discardPile.add(c)
            if (c.rank != 8) break
        }
        currentIndex = 0; justDrew = false
        phase = CrazyEightsPhase.PLAYING
        lastEvent = "${current!!.name} starts"
    }

    fun canPlay(c: Card): Boolean {
        val top = topCard ?: return true
        if (c.rank == 8) return true
        if (c.suit == activeOrTopSuit) return true
        if (c.rank == top.rank) return true
        return false
    }

    /** Returns null on success, an error message on failure. */
    fun playCard(playerId: String, card: Card, declaredSuit: Suit? = null): String? {
        if (phase != CrazyEightsPhase.PLAYING) return "not playing"
        val p = players[playerId] ?: return "unknown player"
        if (p.id != current!!.id) return "not your turn"
        if (!p.hand.contains(card)) return "card not in hand"
        if (!canPlay(card)) return "card does not match"
        if (card.rank == 8 && declaredSuit == null) return "must declare a suit"
        p.hand.remove(card)
        discardPile.add(card)
        stalePasses = 0
        activeSuit = if (card.rank == 8) declaredSuit else null
        justDrew = false
        if (options.queenReverses && card.rank == 12 && order.size > 2) direction = -direction
        val skipNext = options.jackSkips && card.rank == 11 && order.size > 2
        lastEvent = if (card.rank == 8) "${p.name} played $card → ${declaredSuit!!.glyph}"
                    else "${p.name} played $card"
        if (p.hand.isEmpty()) {
            phase = CrazyEightsPhase.GAME_OVER; winnerId = p.id; return null
        }
        advanceTurn()
        if (skipNext) advanceTurn()
        if (options.twosDrawTwo && card.rank == 2 && order.size >= 2) {
            val victim = current!!
            repeat(2) {
                if (drawPile.isEmpty()) reshuffle()
                if (drawPile.isNotEmpty()) victim.hand.add(drawPile.removeAt(drawPile.size - 1))
            }
            lastEvent = "${victim.name} draws two and is skipped"
            advanceTurn()
        }
        return null
    }

    fun drawOne(playerId: String): Card? {
        if (phase != CrazyEightsPhase.PLAYING) return null
        val p = players[playerId] ?: return null
        if (p.id != current!!.id) return null
        if (drawPile.isEmpty()) reshuffle()
        if (drawPile.isEmpty()) {
            stalePasses += 1
            if (stalePasses >= order.size) {
                phase = CrazyEightsPhase.GAME_OVER
                var best = order[0]
                for (pid in order) if (players[pid]!!.hand.size < players[best]!!.hand.size) best = pid
                winnerId = best
                lastEvent = "Stalemate — ${players[best]!!.name} wins with the fewest cards"
                return null
            }
            advanceTurn(); return null
        }
        val c = drawPile.removeAt(drawPile.size - 1)
        p.hand.add(c)
        stalePasses = 0
        lastEvent = "${p.name} drew a card"
        if (canPlay(c)) { justDrew = true; return c }
        justDrew = false; advanceTurn(); return c
    }

    fun passAfterDraw(playerId: String) {
        if (phase != CrazyEightsPhase.PLAYING) return
        val p = players[playerId] ?: return
        if (p.id != current!!.id || !justDrew) return
        justDrew = false
        lastEvent = "${p.name} passed"
        advanceTurn()
    }

    private fun advanceTurn() {
        val n = order.size
        currentIndex = ((currentIndex + direction) % n + n) % n
        justDrew = false
    }

    private fun reshuffle() {
        if (discardPile.size <= 1) return
        val top = discardPile.removeAt(discardPile.size - 1)
        drawPile.addAll(discardPile.shuffled(rng))
        discardPile.clear(); discardPile.add(top)
    }

    fun reset() {
        phase = CrazyEightsPhase.LOBBY
        drawPile.clear(); discardPile.clear()
        activeSuit = null; currentIndex = 0; justDrew = false; direction = 1
        stalePasses = 0
        winnerId = null; lastEvent = null
        for (p in players.values) p.hand.clear()
    }
}
